### ISSUE TITLE
filter() type not properly derived from overriden __iter__()

### ISSUE THREAD

**Issue by Fatsie** [python/mypy#10485]
This seems to be a problem shared between mypy and Pylance.
See description in Pylance bug report: https://github.com/microsoft/pylance-release/issues/1295#issue-892717386
mypy version used is 0.812

**Comment by erictraut** [comment]
Here's a minimal repro of this problem.

```python
class X(Mapping[str, int], Sequence[int]):
    ...
```

This should generate an error because `Mapping[str, int]` is a subtype of `Sequence[str]` which is incompatible with `Sequence[int]`.
